# -*- coding: utf-8 -*-
from qgis.PyQt import uic
from qgis.PyQt.QtWidgets import QDialog, QFileDialog, QMessageBox
from qgis.PyQt.QtCore import QObject, pyqtSignal
from qgis.core import QgsTask, QgsMessageLog, Qgis, QgsApplication
from pathlib import Path
from .gpkg2fusion_tool import fusionar_vectores
import os

# Cargar el UI (aseg√∫rate de que el nombre del archivo .ui sea correcto)
FORM_CLASS, _ = uic.loadUiType(os.path.join(
    os.path.dirname(__file__), 'gpkg2fusion_dialog.ui'))


class LoggerEmitter(QObject):
    """Objeto que emite se√±ales de log para ser conectadas al QTextEdit en el hilo GUI."""
    signal = pyqtSignal(str)


class Gpkg2FusionDialog(QDialog, FORM_CLASS):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setupUi(self)

        # Emisor de logs (se√±al thread-safe)
        self._log_emitter = LoggerEmitter()
        self._log_emitter.signal.connect(self._append_log_threadsafe)

        # Conectar botones
        self.inputBrowseButton.clicked.connect(self.select_input_folder)
        self.outputBrowseButton.clicked.connect(self.select_output_file)
        self.runButton.clicked.connect(self.run_fusion)
        self.cancelButton.clicked.connect(self.cancel_task)

        self.task = None
        self.task_active = False

        # Mensaje inicial
        self.logTextEdit.append("üì¶ Herramienta de fusi√≥n de GPKG")

    def _append_log_threadsafe(self, msg: str):
        """Append al QTextEdit en hilo GUI (conectado a la se√±al)."""
        # dejamos el append simple; la se√±al garantiza ejecuci√≥n en hilo GUI
        self.logTextEdit.append(msg)

    def select_input_folder(self):
        folder = QFileDialog.getExistingDirectory(self, "Seleccionar carpeta de entrada")
        if folder:
            self.inputFolderLineEdit.setText(folder)

    def select_output_file(self):
        path, _ = QFileDialog.getSaveFileName(
            self,
            "Seleccionar archivo GPKG de salida",
            "",
            "GeoPackage (*.gpkg);;Todos los archivos (*)"
        )
        if path and not path.lower().endswith(".gpkg"):
            path += ".gpkg"
        if path:
            self.outputFileLineEdit.setText(path)

    def run_fusion(self):
        input_path = Path(self.inputFolderLineEdit.text().strip())
        output_path = Path(self.outputFileLineEdit.text().strip())
        epsg_text = self.epsgLineEdit.text().strip()

       # Validaciones
        input_path_text = str(input_path).strip()
        if not input_path_text:
            QMessageBox.warning(self, "Error", "Debe seleccionar una carpeta de entrada.")
            return

        if not input_path.exists() or not input_path.is_dir():
            QMessageBox.warning(self, "Error", "Debe seleccionar una carpeta de entrada v√°lida.")
            return

        output_path_text = str(output_path).strip()
        if not output_path_text:
            QMessageBox.warning(self, "Error", "Debe seleccionar un archivo de salida (.gpkg).")
            return

        if not output_path.parent.exists():
            try:
                output_path.parent.mkdir(parents=True, exist_ok=True)
            except Exception:
                QMessageBox.warning(self, "Error", "Ruta de salida no v√°lida.")
                return

        epsg = int(epsg_text) if epsg_text.isdigit() else None

        # Limpiar log
        self.logTextEdit.clear()
        self.logTextEdit.append("üì¶ Iniciando proceso de fusi√≥n...\n")

        # Crear tarea y agregar al task manager
        self.task = GpkgToFusionTask(
            input_path,
            output_path,
            epsg,
            log_emitter=self._log_emitter,
            dialog=self
        )
        self.task_active = True
        self.runButton.setEnabled(False)
        QgsApplication.taskManager().addTask(self.task)

    def cancel_task(self):
        if self.task_active and self.task:
            self.task.cancel()
            # mostrar mensaje inmediato en GUI
            self.logTextEdit.append("‚èπ Cancelando tarea...")

# ----------------------------------------------------
class GpkgToFusionTask(QgsTask):
    def __init__(self, input_path, output_path, epsg, log_emitter: LoggerEmitter, dialog):
        super().__init__("Fusi√≥n de GPKG")
        self.input_path = input_path
        self.output_path = output_path
        self.epsg = epsg
        self.log_emitter = log_emitter
        self.dialog = dialog
        self.cancelled_flag = False

    def cancel(self):
        # marcar cancelaci√≥n (QgsTask.cancel() tambi√©n hace trabajo interno)
        self.cancelled_flag = True
        return super().cancel()

    def run(self):
        # Callbacks usados por gpkg2fusion_tool
        def cancel_cb():
            return self.cancelled_flag

        def log_cb(msg):
            # registrar en QGIS message log inmediatamente
            QgsMessageLog.logMessage(msg, "GPKG Tools", Qgis.Info)
            # emitir por la se√±al para que llegue al QTextEdit en hilo GUI
            if self.log_emitter:
                self.log_emitter.signal.emit(msg)

        try:
            # Ejecutar la fusi√≥n (esta funci√≥n comprobar√° cancel_cb y llamar√° a log_cb)
            fusionar_vectores(
                self.input_path,
                self.output_path,
                epsg_destino=self.epsg,
                log_cb=log_cb,
                cancel_cb=cancel_cb
            )
        except Exception as e:
            log_cb(f"‚ùå Error durante la fusi√≥n: {e}")

        return True

    def finished(self, result):
        # Marcar tarea como inactiva en el di√°logo y reactivar botones
        if self.dialog:
            self.dialog.task_active = False
            if hasattr(self.dialog, "runButton"):
                self.dialog.runButton.setEnabled(True)
        # Mensaje final en GUI
        if self.log_emitter:
            self.log_emitter.signal.emit("‚úÖ Tarea finalizada.")
